### PR TITLE
feat(app): integrate the block-sdk

### PR DESCRIPTION
**Pull request by nivasan1**
### In This PR 
- I integrate the block-sdk
    - I integrate the mev, free, default lanes (ordered accordingly) into the `LaneMempool` which will be used as the app's `Mempool`
    - I integrate the Prepare / ProcessProposal handlers and custom CheckTx handler
    - Modules
        - I integrate the `x/auction` module, and the corresponding `AnteHandler` for the auction module
    - I integrate the `ante.IgnoreDecorator` with the `freeLane`,  and the `DeductFees` ante-handler: this will skip the `DeductFees` ante-handler for any txs that originate from the `freeLane`
### Testing
- We've tested this integration against our own set of load-tests. These tests, spin up a network of 5 nodes (3 val, 1 seed, 1 full). Perform various perturbations against them (killing, restarting, disconnecting from the network, etc.), and periodically send a load of 350 tx/sec for 90s. The output from one of our trials is [here](https://snapshots.raintank.io/dashboard/snapshot/y23ErIIGdQEL8AXCbFdv4grb2hxldbMz?orgId=2), happy to go through these in more depth / review the methodology.
- In order for the e2e tests to work, I had to modify the `ante_no_seq` ante-handlers to ignore signature verification, while incrementing sequence #s, I leave a comment in the section describing why.
### Let's Discuss
- I chose to use stable defaults [here](https://github.com/skip-mev/osmosis/pull/1/files#diff-4aaa0025728f832e685e0651c0febf270306c9150833063c1c9a023c02bfa14aR15) (tx-capacity of each lane) and [here](https://github.com/skip-mev/osmosis/pull/1/files#diff-4aaa0025728f832e685e0651c0febf270306c9150833063c1c9a023c02bfa14aR26) (block-space allocated to each lane), let me know if any of the parameters are confusing or should be changed
- regarding the auction-module
    -  I imagine that the full integration requires the v23 upgrade-handler (store-migration, genesis-state / parameters, etc.), I am happy to implement those as a separate PR if that is desired
    - `Auction` module's parameters are [here](https://github.com/skip-mev/block-sdk/blob/31def89bc1fa8e3460a11b53030b6d965da860f4/x/auction/types/genesis.pb.go#L76), take a look at these or we can go ahead w/ the [defaults](https://github.com/skip-mev/block-sdk/blob/31def89bc1fa8e3460a11b53030b6d965da860f4/x/auction/types/params.go#L11), some things to consider
       - What denomination would we like the auction to accept (there can only be a single denom accepted)
       - What would be the desired max bundle size (maximal # of txs that a bid can include)
       - What address would we like the auction revenue to be sent to? How much of it should we give to proposers?
- re the free-lane: We currently are matching the `distributiontypes.WithdrawDelegatorRewards` tx, the process of adding more txs to this lane is simple, we just need to identify which types we want!

### CODE REVIEW

File: app/ante_no_seq.go [hunk 1]
```diff
@@ -59,12 +74,21 @@ func NewAnteHandler(
...
+			deductFeeDecorator,
+			blockSDKParams.freeLane,
+		),
 		ante.NewSetPubKeyDecorator(ak), // SetPubKeyDecorator must be called before all signature verification decorators
 		ante.NewValidateSigCountDecorator(ak),
 		ante.NewSigGasConsumeDecorator(ak, sigGasConsumer),
-		ante.NewSigVerificationDecorator(ak, signModeHandler),
-		// ante.NewIncrementSequenceDecorator(ak),
+		// ante.NewSigVerificationDecorator(ak, signModeHandler) <-- removed this to prevent failures resulting from invalid tx orders in e2e
+		ante.NewIncrementSequenceDecorator(ak),
```
Comment (nivasan1) [inline app/ante_no_seq.go:85]: notice: This switch is made, as the priority nonce-mempool (underlying mempool used by the block-sdk) requires that all txs have a unique sequence number. As such, we remove the signature verification ante-handler (to ignore the sequence # check in e2e txs), but still increment the seq. to get around over-writing txs in the mempool

File: app/app.go [hunk 2]
```diff
@@ -352,6 +359,21 @@ func NewOsmosisApp(
...
 	app.sm.RegisterStoreDecoders()
 
+	// initialize lanes + mempool
+	mevLane, freeLane, defaultLane := CreateLanes(app, txConfig)
+
+	// create the mempool
+	lanedMempool, err := block.NewLanedMempool(
+		app.Logger(),
+		[]block.Lane{mevLane, freeLane, defaultLane},
+	)
```
Comment (p0mvn) [inline app/app.go:369]: How does `LanedMempool` `Select` transactions from lanes given that `nil` is returned?
https://github.com/skip-mev/block-sdk/blob/31def89bc1fa8e3460a11b53030b6d965da860f4/block/mempool.go#L105-L113

File: app/lanes.go [hunk 3]
```diff
@@ -0,0 +1,118 @@
...
+	mevlane "github.com/skip-mev/block-sdk/lanes/mev"
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	distrtypes "github.com/cosmos/cosmos-sdk/x/distribution/types"
+)
+
+const (
+	maxTxPerMEVLane = 500 // this is the maximum # of bids that will be held in the app-side in-memory mempool
+	maxTxPerFreeLane = 1000 // this is the maximum # of free-txs that will be held in the app-side in-memory mempool
+	maxTxPerDefaultLane = 3000 // all other txs
+)
```
[the remark this answers is not in the document]
Comment (nivasan1) [reply app/lanes.go:20]: Currently, the default for app-side mempools is the `NopMempool`. This means that the max-txs held in the app-side mempool is 0, however, the max-txs for comet is 5k (default). The block-sdk's CheckTx handler ensures that comet / app-side mempools are always at parity. So the mempool size will be 4.5k

File: app/lanes.go [hunk 4]
```diff
@@ -0,0 +1,118 @@
...
+	return func(ctx sdk.Context, tx sdk.Tx) bool {
+		for _, msg := range tx.GetMsgs() {
+			if _, ok := msg.(*distrtypes.MsgWithdrawDelegatorReward); ok {
+				return true
+			}
+		}
+
+		return false
+	}
+}
```
[the remark this answers is not in the document]
Comment (nivasan1) [reply app/lanes.go:118]: We proposed this under direction we received from a skip x osmosis weekly chat, happy to change this further

File: app/app.go [hunk 5]
```diff
@@ -352,6 +359,21 @@ func NewOsmosisApp(
 
 	app.sm.RegisterStoreDecoders()
 
+	// initialize lanes + mempool
+	mevLane, freeLane, defaultLane := CreateLanes(app, txConfig)
```
[the remark this answers is not in the document]
Comment (nivasan1) [reply app/app.go:363]: There are no problems, the integration could be made much more lightweight if desired

File: app/app.go [hunk 6]
```diff
@@ -368,8 +390,53 @@ func NewOsmosisApp(
...
+			txConfig: txConfig,
+		},
+	)
+
+	// update ante-handlers on lanes
+	opt := []base.LaneOption{
+		base.WithAnteHandler(anteHandler),
+	}
+	mevLane.WithOptions(opt...)
+	defaultLane.WithOptions(opt...)
```
[the remark this answers is not in the document]
Comment (nivasan1) [reply app/app.go:406]: Yes, [this function](https://github.com/osmosis-labs/osmosis/pull/7339/files#diff-4aaa0025728f832e685e0651c0febf270306c9150833063c1c9a023c02bfa14aR84), the `MatchHandler` allows integrators to define any criterion they'd like for what txs belong in any lane

File: app/app.go [hunk 7]
```diff
@@ -352,6 +359,21 @@ func NewOsmosisApp(
...
+	// create the mempool
+	lanedMempool, err := block.NewLanedMempool(
+		app.Logger(),
+		[]block.Lane{mevLane, freeLane, defaultLane},
+	)
+	if err != nil {
+		panic(err)
+	}
+	// set the mempool
+	app.SetMempool(lanedMempool)
```
[the remark this answers is not in the document]
Comment (nivasan1) [reply app/app.go:374]: So all that would be needed is an impl of [this interface](https://github.com/skip-mev/block-sdk/blob/acff9d0edf3b4a710dcc67f27ab3168a1b1cdfda/block/lane.go#L17), and on the lanes you'd like to change the backing mempool, you'd j `lane.WithMempool(mempl)`

File: app/app.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (nivasan1) [reply app/app.go:406]: The ordering of txs is determined by the underlying mempool + PrepareLane handlers, each lane may have a different ordering criterion. For the free / default lanes, txs are ordered by sender nonce priority in accordance with the `PriorityNonce` mempool.

File: app/app.go (hunk 2 above)
Comment (nivasan1) [reply app/app.go:369]: Great question! See [here](https://github.com/skip-mev/block-sdk/blob/31def89bc1fa8e3460a11b53030b6d965da860f4/abci/abci.go#L48) for the Prepare / ProcessProposal handlers. Essentially, how blocks are built is that as opposed to calling `app.Mempool.Select()` (which is a no-op), we iterate through `app.LanedMempool.Lanes()`, and call `lane.Select()` on each of the sub-lanes. From that lane we select up to the max-block-space / max-txs allocated for that lane. Does that make sense?

File: app/lanes.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (czarcas7ic) [reply app/lanes.go:118]: We probably need to ensure the amount being withdrawn hits a certain threshold to prevent this being taken advantage of right?

**Comment by PaddyMc** [comment]
https://github.com/osmosis-labs/osmosis/pull/7387

:arrow_up: upgrade handler has been generated, please feel free to add the relevant migrations :smile:

File: app/lanes.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (p0mvn) [reply app/lanes.go:20]: I looked at the mainnet graphs, and we seem to be far from reaching the mempool size limit. Was this what guided lowering the overall limit from 5K to 4.5K?

File: app/lanes.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (p0mvn) [reply app/lanes.go:118]: Previously discussed this with @nivasan1 async.

From @nivasan1 
> So, looking into the MsgWithdrawDelegatorReward filter, we'd need to introspect state to determine what the rewards to be withdrawn are (this is not a parameter of the message), this would make this filter break key assumptions of the MatchHandler in that it's execution could differ between ingress and insertion into a Proposal. Would this option still be interesting? This may be possible if we were to Nop ProcessProposal

My reply:
> Hmm that makes sense. In the end, we would still like to move towards the non-Nop ProcessProposal. So if we were to implement this, it would have to be redesigned.
In that case, I think it is fine to leave the withdrawal as is. That is, without the conditional logic based on reward value and simply go to free lane.
I think that the original proposal for making MsgWithdawDelegatorReward in such a way was to prevent spam. However, spam is still capped at the lane size. In addition, a spammer first needs to delegate with a non-zero tx fee to be able to spam withdrawals that are limited by the lane size.

TLDR; we should be fine leaving this as is
